src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,189 @@
+        private readonly ConcurrentSet<ISymbol> _foundDefinitions = new ConcurrentSet<ISymbol>();
+
+        /// <remarks>
+        /// _referencesCount is read and written from multiple threads.
+        /// so all read and write should be through a lock around _gate.
+        /// </remarks>
+        private int _referencesCount;
+
+        /// <remarks>
+        /// If the cap is 0, then there is no cap.
+        /// </remarks>
+        public int SearchCap { get; }
+
+        public CodeLensFindReferencesProgress(
+            ISymbol queriedDefinition,
+            SyntaxNode queriedNode,
+            CancellationTokenSource cancellationTokenSource,
+            int searchCap)
+        {
+            _queriedSymbol = queriedDefinition;
+            _queriedNode = queriedNode;
+            _cancellationTokenSource = cancellationTokenSource;
+
+            SearchCap = searchCap;
+        }
+
+        public bool SearchCapReached
+        {
+            get
+            {
+                if (SearchCap == 0)
+                {
+                    return false;
+                }
+
+                lock (_gate)
+                {
+                    return _referencesCount > SearchCap;
+                }
+            }
+        }
+
+        public int ReferencesCount
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _referencesCount;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns partial symbol locations whose node does not match the given syntaxNode
+        /// </summary>
+        /// <param name="symbol">Symbol whose locations are queried</param>
+        /// <param name="syntaxNode">Syntax node to compare against to exclude location - actual location being queried</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Partial locations</returns>
+        private static IEnumerable<Location> GetPartialLocations(ISymbol symbol, SyntaxNode syntaxNode, CancellationToken cancellationToken)
+        {
+            // Returns nodes from source not equal to actual location
+            IEnumerable<SyntaxNode> syntaxNodes = from syntaxReference in symbol.DeclaringSyntaxReferences
+                                                  let candidateSyntaxNode = syntaxReference.GetSyntax(cancellationToken)
+                                                  where !(syntaxNode.Span == candidateSyntaxNode.Span &&
+                                                          syntaxNode.SyntaxTree.FilePath.Equals(candidateSyntaxNode.SyntaxTree.FilePath, StringComparison.OrdinalIgnoreCase))
+                                                  select candidateSyntaxNode;
+
+            // This matches the definition locations to syntax references, ignores metadata locations
+            IEnumerable<Location> partialLocations = from currentSyntaxNode in syntaxNodes
```

Comment:
is this to convert location syntax reference is pointing to to where location is pointing to?